File: daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.daffodil.runtime1.layers.api

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.charset.Charset
import java.util.Optional

import org.apache.daffodil.runtime1.api.DFDLPrimType
import org.apache.daffodil.runtime1.layers.LayerException

// TODO: Convert this whole file to Java ??

/**
 * Descriptor of the DFDL variables used by the layer.
 *
 * The names and types must match the dfdl:defineVariable definitions used in
 * an included/imported DFDL component schema which provides the layer definition for use
 * by other schemas.
 *
 * The first such variable is sometimes distinguished as the variable written to
 * as the single unique value of the layer, such as when the layer computes a
 * checksum. This is, however, just a convention used by various checksum layers
 * and classes/traits that support writing checksum layers.
 *
 * @param prefix preferred namespace prefix for the namespace of the variables
 * @param namespace namespace of the variables
 * @param variables list of pairs, each is the name (an NCName, that is without any
 *                  namespace prefix) and type of a variable
 */
final case class LayerVariables(
  prefix: String,
  namespace: String,
  variables: java.util.List[(String, DFDLPrimType)],
)

/**
 * This is the primary API class for writing layers.
 *
 * All layers are derived from this class, and must have no-args default constructors.
 *
 * Derived classes will be dynamically loaded by Java's SPI system.
 * The names of concrete classes derived from Layer are listed in a resources/M.services file
 * so that they can be found and dynamically loaded.
 *
 * The SPI creates an instance the class of which is used as a factory to create the
 * instances actually used by the Daffodil runtime. Compilation of the static information about
 * the layer occurs only once and is then shared by all runtime instances.
 *
 * Instances of derived layer classes can be stateful. They are private to threads, and each time a layer
 * is encountered during parse/unparse, an instance is created for that situation.
 *
 * All the static information about the layer is provided in the arguments.
 *
 * The rest of the Layer class implements the
 * layer decode/encode logic, which is done as part of deriving one's Layer class from the
 * Layer base class.
 *
 * About variables: Layer logic may read and write variables. Variables being read are parameters to
 * the layer algorithm. Variables being written are outputs (such as checksums) from the layer algorithm.
 * Variables being written must be undefined, since variables in DFDL are single-assignment.
 * Variables being read must be defined before being read by the layer, and this is true for both
 * parsing and unparsing. When unparsing, variables being read cannot be forward-referencing to parts
 * of the DFDL infoset that have not yet been unparsed.
 *
 * @param layerName the name that will appear in the DFDL schema to identify the layer
 * @param supportedLayerLengthKinds list of the layer length kinds the layer supports
 * @param supportedLayerLengthUnits list of the layer length units the layer supports
 * @param isRequiredLayerEncoding true if the layer is textual and so needs the layerEncoding property
 * @param optLayerVariables a LayerVariables structure describing the variables the layer accesses
 */
abstract class Layer(
  val layerName: String,
  val supportedLayerLengthKinds: java.util.List[JLayerLengthKind],
  val supportedLayerLengthUnits: java.util.List[JLayerLengthUnits],
  val isRequiredLayerEncoding: Boolean,
  val optLayerVariables: Optional[LayerVariables],
) {

  final def name(): String =
    layerName // name() method with empty args is required by SPI loader

  /**
   * Called exactly once when the schema is compiled to do extra checking that the layer is being used properly.
   * The thrown exception becomes a SchemaDefinitionError at schema compile time.
   *
   * Example checks are:
   * - layerEncoding is constant and is a single-byte charset
   * - layerLength, if constant, is within a maximum value range
   * - layerBoundaryMark string, if constant, is not too long and contains only allowed characters.
   * These things can be required to be constant by this check, or it can check their values for legality
   * if they happen to be constant. Since these are runtime-valued properties (can be expressions), then if the
   * layer allowed that, they must also be checked at runtime.
   *
   * You don't have to check that the variables are defined and declared in matching manner, that happens automatically.
   *
   * @throws LayerException on any error.
   *
   */
  @throws[LayerException]
  def check(layerPropertyInfo: LayerPropertyInfo): Unit = { /* nothing */ }
  // TODO: figure out what to throw. LayerCompilerError ??

  @throws[IOException]
  def wrapLayerDecoder(jis: InputStream, lr: LayerRuntime): InputStream

  @throws[IOException]
  def wrapLayerEncoder(jos: OutputStream, lr: LayerRuntime): OutputStream

  /**
   * Override if you need to customize layer length limiting behavior.
   * You will not override if you use standard explicit, implicit, or boundary-mark behaviors.
   *
   * @return a LayerLimiter or null to indicate the standard layer limiter implementation should be used.
   */
  def layerLimiter(layerPropertyInfo: LayerPropertyInfo): Optional[LayerLimiter] =
    Optional.empty()
}

/**
 * Layers that use the standard layerLengthKinds will not need to use this.
 *
 * But if they want to implement different behaviors for how the length is determined
 * then the layer must implement a LayerLimiter.
 *
 * Implementations of this trait must have a public zero-arg constructor which will be
 * used to allocate a per-thread per-layer instance of the class.
 *
 * That is, implementations of the class are free to be stateful, as their state
 * will be private to just that single invocation. 
 */
trait LayerLimiter {
  def wrapLayerLimitingInputStream(jis: InputStream, lr: LayerRuntime): InputStream
  def wrapLayerLimitingOutputStream(jos: OutputStream, lr: LayerRuntime): OutputStream
}

/**
 * Runtime information and stateful services available to the layer when
 * encoding/decoding the layer data to/from the input/output stream.
 *
 * Provides the ability to cause runtime processing errors, which can cause backtracking
 * when parsing, but are fatal when unparsing.
 *
 * Also provides the ability to cause runtime schema definition errors, which are always
 * fatal.
 *
 * This object contains the processor state, but hidden behind an API so that only relevant
 * aspects of the processor state are visible. 
 */
trait LayerRuntime extends LayerPropertyInfo {

  def layerLength(): Long

  def layerBoundaryMark(): String

  /**
   * @return the value of the layerEncoding property converted to a java.nio.charset.Charset.
   *         If defined as an expression, the expression is evaluated and the
   *         string it produces interpreted as the name of a charset encoding. If the layerEncoding property
   *         is not defined it is an error in the layer implementation reported as a schema definition error.
   *         If the layerEncoding string value cannot be converted to a charset, then it is a processing error.
   */
  def layerCharset: Charset

  /**
   * At runtime, for speed reasons, variables are accessed by index which is the zero-based position in
   * the LayerVariables variables list.
   * @param index zero-based index selecting the variable structure
   * @return a LayerVariable which allows setting and getting the variable value.
   */
  def variable(index: Int): LayerVariable

  /**
   * Used when parsing to cause a runtime ParseError from layer code. Parse errors can cause backtracking.
   * Used when unparsing to cause a runtime UnparseError (which is generally fatal) from layer code.
   *
   * @param msg text that will be included in any diagnostic output.
   * @return This method does not return. Control is thrown to the Daffodil runtime.
   */
  def processingError(msg: String): Nothing

  /**
   * Used when parsing to cause a runtime ParseError from layer code. Parse errors can cause backtracking.
   * Used when unparsing to cause a runtime UnparseError (which is generally fatal) from layer code.
   *
   * @param cause a throwable that is the cause of the error
   * @return This method does not return. Control is thrown to the Daffodil runtime.
   */
  def processingError(cause: Throwable): Nothing

  /**
   * Used when parsing or unparsing to indicate that the schema is not meaningful.
   *
   * An example of this is if a runtime-valued property (defined by an expression) such as layerEncoding
   * does not have a suitable value for the layer to use. For example, it may evaluate to the name of a valid
   * encoding, but the encoding might not be supported; some layers can only work with Single-Byte Character Sets.
   *
   * This method does not return. Control is thrown to the Daffodil runtime.
   * @param msg text that will be displayed as part of the diagnostic information.
   * @param args arguments substituted into the msg as in String.format
   */
  def runtimeSchemaDefinitionError(msg: String, args: AnyRef*): Nothing

  /**
   * Used when parsing or unparsing to indicate that the schema is not meaningful.
   *
   * An example of this is if a runtime-valued property (defined by an expression) such as layerEncoding
   * does not have a suitable value for the layer to use. For example, it may evaluate to the name of a valid
   * encoding, but the encoding might not be supported; some layers can only work with Single-Byte Character Sets.
   *
   * This method does not return. Control is thrown to the Daffodil runtime.
   * @param cause Exception object that captures the cause of the problem.
   */
  def runtimeSchemaDefinitionError(cause: Throwable): Nothing

  /**
   * @param variable the variable to get
   * @return the value of the variable as a string. If the variable has no value it is a processing error.
   */
  def getString(variable: LayerVariable): String

  /**
   * Sets the value of the variable. Note that variables in DFDL are _single assignment_.
   * So if a layer sets a variable more than once it is an error in the layer code
   * which is reported as a runtime schema definition error (fatal).
   * @param variable the variable to set
   * @param s the string value to which the variable will be set
   */
  def setString(variable: LayerVariable, s: String): Unit

  /**
   * @param variable the variable to get
   * @return the value of the variable as a signed Int. The variable must be of integer (or unsigned integer) type, and if it
   *         is not it is an error in the layer code which is reported as a runtime schema definition error. (This may
   *         mean the layer code definition of the variable, and the DFDL schema's dfdl:defineVariable disagree on the
   *         type.) No attempt is made to convert strings to integers. If the variable has no value it is a processing error.
   *         If the value is greater than Int.MaxValue then it is a processing error.
   */
  def getInt(variable: LayerVariable): Int

  /**
   * Sets the value of the variable. Note that variables in DFDL are _single assignment_.
   * So if a layer sets a variable more than once it is an error in the layer code
   * which is reported as a runtime schema definition error (fatal). The variable must be of
   * integer (or unsigned integer) type. No attempt is made to convert the argument into a string
   * if the variable type is not an integer type. This type error is reported as a runtime schema definition error.
   *
   * @param variable the variable to set
   * @param v the value to assign. If the type of the variable is unsignedInt, and this parameter is negative, it is
   *          a processing error.
   */
  def setInt(variable: LayerVariable, v: Int): Unit

}

/**
 * Variable values are obtained or set using LayerVariable objects.
 */
trait LayerVariable

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.daffodil.runtime1.layers.api;

import org.apache.daffodil.runtime1.layers.LayerRuntime;
import org.apache.daffodil.runtime1.layers.LayerUtils;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * This is the primary API class for writing layers.
 * <p/>
 * All layers are derived from this class, and must have no-args default constructors.
 * <p/>
 * Derived classes will be dynamically loaded by Java's SPI system.
 * The names of concrete classes derived from Layer are listed in a resources/META-INF/services file
 * so that they can be found and dynamically loaded.
 * <p/>
 * The SPI creates an instance the class by calling a default (no-arg) constructor, which should be
 * the only constructor.
 * <p/>
 * Instances of derived layer classes can be stateful. They are private to threads, and each time a layer
 * is encountered during parse/unparse, an instance is created for that situation.
 * <p/>
 * Layer instances should not share mutable state (such as via singleton objects)
 * <p/>
 * The rest of the Layer class implements the
 * layer decode/encode logic, which is done as part of deriving one's Layer class from the
 * Layer base class.
 * <p/>
 * About variables: Layer logic may read and write DFDL variables.
 * <p/>
 * Every DFDL Variable in the layer's targetNamespace is used either at the start of the
 * layer algorithm as a parameter to the layer or at the end of the layer algorithm it is
 * assigned as a return value (such as a checksum) from the layer.
 * <p/>
 * Variables being written must be undefined, since variables in DFDL are single-assignment.
 * <p/>
 * Variables being read must be defined before being read by the layer, and this is true for both
 * parsing and unparsing. When unparsing, variables being read cannot be forward-referencing to parts
 * of the DFDL infoset that have not yet been unparsed.
 * <p/>
 * A layer that wants to read parameters declares special setter named 'setLayerVariableParameters'
 * which has args such that each has a name and type that match a corresponding
 * dfdl:defineVariable in the layer's namespace.
 * <p/>
 * A layer that wants to return a value after the layer algorithm completes defines a special recognizable
 * getter method. The name of the getter is formed from prefixing the DFDL variable name with the string
 * 'getLayerVariableResult_'. The return type of the getter must match the type of the variable.
 * <p/>
 * For example, a result value getter for a DFDL variable named 'checksum' of type xs:unsignedShort would be:
 * <pre>
 *      int getLayerVariableResult_checksum() {
 *          // returns the value created by the checksum algorithm.
 *      }
 * </pre>
 * <p/>
 */
public abstract class Layer {

  private final String localName;
  private final String targetNamespace;

  private LayerRuntime layerRuntime;

  /**
   * Constructs a new Layer object with the given layer name and namespace.
   *
   * @param localName      the local NCName of the layer. Must be usable as a Java identifier.
   * @param targetNamespace the namespace of the layer. Must obey URI syntax.
   * @throws IllegalArgumentException if arguments are null or do not obey required syntax.
   */
  public Layer(String localName, String targetNamespace) {

    LayerUtils.requireJavaIdCompatible(localName, "layerLocalName");
    LayerUtils.requireURICompatible(targetNamespace, "layerNamespace");

    this.localName = localName;
    this.targetNamespace = targetNamespace;
  }

  /** The spiName of the Layer class.
   * <p/>
   * This method and the string it returns are required by the SPI loader.
   * @return A unique indentifier for the kind of layer. Contains both local and namespace components of the layer's complete name.
   */
  public final String name() { return LayerUtils.spiName(localName, targetNamespace); }

  public final String localName() { return this.localName; }
  public final String namespace() { return this.targetNamespace; }

  /**
   * Called by the execution framework to give the context for reporting errors.
   * @param lr runtime data structure used by the framework
   */
  public final void setLayerRuntime(LayerRuntime lr) {
    this.layerRuntime = lr;
  }
  public final LayerRuntime getLayerRuntime() { return layerRuntime; }

  /**
   * Use to report a processing error.
   * <p/>
   * When parsing a processing error can cause backtracking so that the parse
   * can often recover from the error.
   * <p/>
   * When unparsing a processing error is fatal.
   * @param msg describes the error
   */
  public void processingError(String msg) { layerRuntime.processingError(msg); }

  /**
   * Use to report a processing error.
   * <p/>
   * When parsing a processing error can cause backtracking so that the parse
   * can often recover from the error.
   * <p/>
   * When unparsing a processing error is fatal.
   * @param cause a throwable object that describes the error
   */
  public void processingError(Throwable cause) { layerRuntime.processingError(cause); }
  /**
   * Use to report a runtime schema definition error.
   * <p/>
   * This indicates that the layer is unable to
   * work meaningfully because of the way it is configured. The schema itself is not well defined due to
   * the way the layer is configured.
   * <p/>
   * This error type is always fatal whether parsing or unparsing.
   * <p/>
   * @param msg describes the error
   */
  public void runtimeSchemaDefinitionError(String msg) { layerRuntime.runtimeSchemaDefinitionError(msg); }

  /**
   * Use to report a runtime schema definition error.
   * <p/>
   * This indicates that the layer is unable to
   * work meaningfully because of the way it is configured. The schema itself is not well defined due to
   * the way the layer is configured.
   * <p/>
   * This error type is always fatal whether parsing or unparsing.
   * <p/>
   * @param cause a throwable object that describes the error
   */
  public void runtimeSchemaDefinitionError(Throwable cause) { layerRuntime.runtimeSchemaDefinitionError(cause); }

  /**
   * Wraps a layer input interpreter around an input stream, using the provided LayerRuntimeFoo for runtime information and stateful services.
   *
   * @param jis The input stream to be wrapped.
   * @return An input stream with the layer wrapped around it.
   */
  public abstract InputStream wrapLayerInput(InputStream jis) throws Exception;

  /**
   * Wraps a layer output interpreter around an output stream, using the provided LayerRuntimeFoo for runtime information and stateful services.
   *
   * @param jos The output stream to be wrapped.
   * @return An output stream with the layer wrapped around it.
   */
  public abstract OutputStream wrapLayerOutput(OutputStream jos) throws Exception;

  private ArrayList<Class<? extends Exception>> peExceptions = new ArrayList<>();

  /**
   * Adds an exception class to the list of exceptions that will be automatically converted
   * into processing errors.
   * <p/>
   * The purpose of this is to allow one to use java/scala libraries that may throw
   * exceptions when encountering bad data. Such exceptions should be translated into
   * processing errors, which will allow the parser to backtrack and try other alternatives
   * which may work for that data.
   * <p/>
   * When considering whether a thrown Exception is to be converted to a processing error
   * RuntimeException classes are handled separately from Exception classes.
   * Hence calling
   * <pre>
   *     setProcessingErrorException(Exception.class);
   * </pre>
   * will NOT cause all RuntimeExceptions to also be converted into processing errors.
   * It will, however, cause all classes derived from Exception that are NOT RuntimeExceptions
   * to be caught and converted into Processing Errors.
   *
   * @param e the exception class to be added to the list of processing error exceptions
   */
  public final void setProcessingErrorException(Class<? extends Exception> e) {
    peExceptions.add(e);
  }

  public final List<Class<? extends Exception>> getProcessingErrorExceptions() {
    return peExceptions;
  }

}